File: happykopiAPI/happykopiAPI/Services/Implementations/OrderService.cs
using Dapper;
using happykopiAPI.Services.Interfaces;
using Microsoft.Data.SqlClient;
using System.Data;
using System.Text.Json;
using happykopiAPI.DTOs.Order.Outgoing_Data;
using happykopiAPI.DTOs.Order.Ingoing_Data;
using happykopiAPI.DTOs.Order.Internal;
using happykopiAPI.DTOs.Product.Outgoing_Data;
using happykopiAPI.DTOs.Product.Internal;
using happykopiAPI.DTOs.Modifier.Internal;
using happykopiAPI.Enums;

namespace happykopiAPI.Services.Implementations
{
    public class OrderService : IOrderService
    {
        private readonly IConfiguration _configuration;
        private readonly INotificationService _notificationService;

        public OrderService(IConfiguration configuration, INotificationService notificationService)
        {
            _configuration = configuration;
            _notificationService = notificationService;
        }

        private IDbConnection CreateConnection() => new SqlConnection(_configuration.GetConnectionString("LocalDB"));

        public async Task<NewOrderResponseDto> CreateOrderAsync(NewOrderRequestDto request)
        {
            using var connection = CreateConnection();

            try
            {
                var orderItemsJson = request.OrderItems.Select(item => new NewOrderItemJsonDto
                {
                    productVariantId = item.ProductVariantId,
                    quantity = item.Quantity,
                    price = item.Price,
                    subtotal = item.Subtotal,
                    modifiers = item.Modifiers.Select(mod => new NewOrderModifierJsonDto
                    {
                        modifierId = mod.ModifierId,
                        quantity = mod.Quantity,
                        price = mod.Price,
                        subtotal = mod.Subtotal
                    }).ToList()
                }).ToList();

                var orderItemsJsonString = JsonSerializer.Serialize(orderItemsJson, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                });

                var parameters = new DynamicParameters();
                parameters.Add("@UserId", request.UserId);
                parameters.Add("@OrderDate", DateTime.Now);
                parameters.Add("@TotalAmount", request.TotalAmount);
                parameters.Add("@Status", request.Status);
                parameters.Add("@PaymentType", request.PaymentType);
                parameters.Add("@AmountPaid", request.AmountPaid);
                parameters.Add("@Change", request.Change);
                parameters.Add("@ReferenceNumber", request.ReferenceNumber);
                parameters.Add("@OrderItemsJson", orderItemsJsonString);
                parameters.Add("@OrderNumber", dbType: DbType.String, direction: ParameterDirection.Output, size: 40);
                parameters.Add("@OrderId", dbType: DbType.Int32, direction: ParameterDirection.Output);

                await connection.ExecuteAsync(
                    "sp_InsertOrder",
                    parameters,
                    commandType: CommandType.StoredProcedure
                );

                var orderId = parameters.Get<int>("@OrderId");
                var orderNumber = parameters.Get<string>("@OrderNumber");

                await _notificationService.NotifyTransactionUpdatedAsync();

                return new NewOrderResponseDto
                {
                    OrderId = orderId,
                    OrderNumber = orderNumber,
                    Status = "SUCCESS",
                    Message = "Order created successfully",
                    OrderDate = DateTime.Now,
                    TotalAmount = request.TotalAmount,
                    AmountPaid = request.AmountPaid,
                    Change = request.Change
                };
            }
            catch (SqlException ex)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new Exception($"Error creating order: {ex.Message}", ex);
            }
        }

        public async Task<IEnumerable<CategoryWithProductCountDto>> GetCategoriesWithProductCountAsync()
        {
            using var connection = CreateConnection();
            return await connection.QueryAsync<CategoryWithProductCountDto>(
                "dbo.sp_GetCategoriesWithProductCount",
                commandType: CommandType.StoredProcedure);
        }

        public async Task<IEnumerable<ProductsWithCategoryDto>> GetProductsWithCategoriesAsync(int categoryId)
        {
            using var connection = CreateConnection();

            return await connection.QueryAsync<ProductsWithCategoryDto>(
                "dbo.sp_GetProductsWithCategoryOf",
                new { CategoryId = categoryId },
                commandType: CommandType.StoredProcedure
            );
        }

        // NEW METHOD: Get Product Availability
        public async Task<ProductAvailabilityResponseDto> GetProductAvailabilityAsync(int? categoryId = null)
        {
            using var connection = CreateConnection();

            var parameters = new { CategoryId = categoryId };

            using var multi = await connection.QueryMultipleAsync(
                "sp_CheckProductAvailability",
                parameters,
                commandType: CommandType.StoredProcedure
            );

            // Read first result set: all products with availability
            var allProducts = (await multi.ReadAsync<ProductAvailabilityFromDbDto>()).ToList();

            // Read second result set: ingredient issues
            var ingredientIssues = (await multi.ReadAsync<IngredientIssueFromDbDto>()).ToList();

            // Group ingredient issues by product
            var issuesByProduct = ingredientIssues
                .GroupBy(i => i.ProductId)
                .ToDictionary(g => g.Key, g => g.ToList());

            // Separate available and unavailable products
            var available = allProducts
                .Where(p => p.IsAvailable)
                .Select(p => new ProductWithAvailabilityDto
                {
                    ProductId = p.ProductId,
                    ProductName = p.ProductName,
                    CategoryName = p.CategoryName,
                    Price = p.Price,
                    ImageUrl = p.ImageUrl,
                    IsAvailable = true
                })
                .ToList();

            var unavailable = allProducts
                .Where(p => !p.IsAvailable)
                .Select(p =>
                {
                    var issues = issuesByProduct.ContainsKey(p.ProductId)
                        ? issuesByProduct[p.ProductId]
                        : new List<IngredientIssueFromDbDto>();

                    var reason = issues.Any() ? issues.First().Reason : "UNKNOWN";

                    return new UnavailableProductDto
                    {
                        ProductId = p.ProductId,
                        ProductName = p.ProductName,
                        CategoryName = p.CategoryName,
                        Price = p.Price,
                        ImageUrl = p.ImageUrl,
                        Reason = reason,
                        Details = issues.Select(i => new IngredientIssueDto
                        {
                            IngredientName = i.IngredientName,
                            Required = i.Required,
                            Available = i.Available,
                            UnitOfMeasure = i.UnitOfMeasure,
                            ExpiredBatchCount = i.ExpiredBatchCount,
                            TotalBatchCount = i.TotalBatchCount
                        }).ToList()
                    };
                })
                .ToList();

            return new ProductAvailabilityResponseDto
            {
                AvailableProducts = available,
                UnavailableProducts = unavailable
            };
        }

        public async Task<IEnumerable<ModifierCountDto>> GetModifierCountByTypeAsync()
        {
            using var connection = CreateConnection();
            var resultsFromDb = await connection.QueryAsync<ModifierCountFromDbDto>(
                "sp_GetModifierCountByType",
                commandType: CommandType.StoredProcedure
            );

            var result = resultsFromDb
                .Select(r => new ModifierCountDto
                {
                    ModifierType = ((ModifierType)r.ModifierType).ToString(),
                    ModifierCount = r.ModifierCount
                })
                .ToList();

            return result;
        }

        public async Task<IEnumerable<OrderModifierSummaryDto>> GetModifiersByTypeAsync(ModifierType modifierType)
        {
            using var connection = CreateConnection();
            var parameters = new { ModifierType = (int)modifierType };
            return await connection.QueryAsync<OrderModifierSummaryDto>("sp_GetModifiersByType", parameters, commandType: CommandType.StoredProcedure);
        }

        public async Task<IEnumerable<OrderModifierSummaryDto>> GetAllModifiersAsync()
        {
            using var connection = CreateConnection();
            return await connection.QueryAsync<OrderModifierSummaryDto>("sp_GetModifiers", commandType: CommandType.StoredProcedure);
        }

        public async Task<IEnumerable<OrderModifierSummaryDto>> GetAvailableModifiersAsync()
        {
            using var connection = CreateConnection();
            return await connection.QueryAsync<OrderModifierSummaryDto>("sp_GetAvailableModifiers", commandType: CommandType.StoredProcedure);
        }

        public async Task<ProductConfigurationResultDto> GetProductConfigurationByIdAsync(int productId)
        {
            using var connection = CreateConnection();
            var parameters = new { p_ProductId = productId };

            var multi = await connection.QueryMultipleAsync(
                "sp_GetProductConfigurationByProductId",
                parameters,
                commandType: CommandType.StoredProcedure
            );

            var variants = (await multi.ReadAsync<OrderVariantDto>()).ToList();
            Console.WriteLine($"[DEBUG] Variants: {variants.Count}");

            // Read ingredients as dynamic to see actual column names
            var ingredientsRaw = (await multi.ReadAsync<dynamic>()).ToList();
            var ingredients = new List<OrderVariantIngredientDto>();

            foreach (var raw in ingredientsRaw)
            {
                var ing = new OrderVariantIngredientDto
                {
                    ProductVariantId = raw.ProductVariantId,
                    StockItemId = raw.StockItemId,
                    StockItemName = raw.StockItemName,
                    QuantityNeeded = raw.QuantityNeeded,
                    UnitOfMeasure = raw.UnitOfMeasure,
                    AvailableStock = raw.AvailableStock
                };

                Console.WriteLine($"[DEBUG] {ing.StockItemName}: Available={ing.AvailableStock}, Needed={ing.QuantityNeeded}");
                ingredients.Add(ing);
            }

            var addOns = (await multi.ReadAsync<OrderVarianAddontDto>())
                .Where(a => a.ModifierId != null && a.ModifierId > 0)
                .ToList();

            var allAvailableAddons = (await multi.ReadAsync<OrderModifierSummaryDto>()).ToList();

            return new ProductConfigurationResultDto
            {
                Variants = variants,
                Ingredients = ingredients,
                AddOns = addOns,
                AllAvailableAddons = allAvailableAddons
            };
        }
    }
}

File: happykopiApp/src/app/modules/pos/services/order.service.ts
// features/order/services/order.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiService } from '../../../core/services/api/api.service';
import { CategoryWithProductCountDto } from '../../../core/dtos/order/category-with-product-count.dto';
import { ProductsWithCategoryDto } from '../../../core/dtos/order/products-with-category.dto';
import { ModifierType } from '../../../core/enums/modifier-type';
import { OrderModifierSummaryDto } from '../../../core/dtos/order/order-modifier-summary.to';
import { HttpParams } from '@angular/common/http';
import { ProductConfigurationResultDto } from '../../../core/dtos/order/product-configuration-result.dto';
import { NewOrderRequestDto } from '../../../core/dtos/order/new-order-request.dto';
import { NewOrderResponseDto } from '../../../core/dtos/order/new-order-response.dto';
import {
  ProductAvailabilityResponseDto,
  UnavailableProductDto,
} from '../../../core/dtos/order/product-availability.dto';

export interface ProductsWithAvailabilityResult {
  products: ProductsWithCategoryDto[];
  unavailableMap: Map<number, UnavailableProductDto>;
}

@Injectable({
  providedIn: 'root',
})
export class OrderService {
  private readonly controller = 'order';

  constructor(private api: ApiService) {}

  getCategories(): Observable<CategoryWithProductCountDto[]> {
    return this.api.get<CategoryWithProductCountDto[]>(`${this.controller}/categories`);
  }

  getAllProducts(categoryId: number | null): Observable<ProductsWithCategoryDto[]> {
    let url = `${this.controller}/products`;
    if (categoryId) {
      url += `?categoryId=${categoryId}`;
    }
    return this.api.get<ProductsWithCategoryDto[]>(url);
  }

  getProductsByCategory(categoryId: number): Observable<ProductsWithCategoryDto[]> {
    return this.api.get<ProductsWithCategoryDto[]>(`${this.controller}/category/${categoryId}`);
  }

  getModifiersByType(type: ModifierType): Observable<OrderModifierSummaryDto[]> {
    const params = new HttpParams().set('modifierType', type.toString());
    return this.api.get<OrderModifierSummaryDto[]>(`${this.controller}/modifiers`, params);
  }

  getProductConfiguration(productId: number): Observable<ProductConfigurationResultDto> {
    return this.api.get<ProductConfigurationResultDto>(
      `${this.controller}/configuration/${productId}`
    );
  }

  getModifiersByIds(modifierIds: number[]): Observable<OrderModifierSummaryDto[]> {
    return this.api.post<OrderModifierSummaryDto[]>(
      `${this.controller}/modifiers/by-ids`,
      modifierIds
    );
  }

  createOrder(orderRequest: NewOrderRequestDto): Observable<NewOrderResponseDto> {
    return this.api.post<NewOrderResponseDto>(this.controller, orderRequest);
  }

  // NEW: Get products with availability info
  getProductsWithAvailability(categoryId?: number): Observable<ProductsWithAvailabilityResult> {
    const params = categoryId
      ? new HttpParams().set('categoryId', categoryId.toString())
      : undefined;

    return this.api
      .get<ProductAvailabilityResponseDto>(`${this.controller}/availability`, params)
      .pipe(
        map((response) => {
          // Map available products
          const availableProducts: ProductsWithCategoryDto[] = response.availableProducts.map(
            (p) => ({
              id: p.productId,
              name: p.productName,
              categoryName: p.categoryName,
              categoryId: 0,
              price: p.price,
              imageUrl: p.imageUrl ?? undefined,
              isAvailable: true,
            })
          );

          // Map unavailable products
          const unavailableProducts: ProductsWithCategoryDto[] = response.unavailableProducts.map(
            (p) => ({
              id: p.productId,
              name: p.productName,
              categoryName: p.categoryName,
              categoryId: 0,
              price: p.price,
              imageUrl: p.imageUrl ?? undefined,
              isAvailable: false,
            })
          );

          // Combine: available first, unavailable at the end
          const allProducts = [...availableProducts, ...unavailableProducts];

          // Create map for unavailable details
          const unavailableMap = new Map<number, UnavailableProductDto>();
          response.unavailableProducts.forEach((p) => {
            unavailableMap.set(p.productId, p);
          });

          return { products: allProducts, unavailableMap };
        })
      );
  }
}